src/DensityEquationTerms.jl

```diff
@@ -0,0 +1,120 @@
+├── timestepper: QuasiAdamsBashforth2TimeStepper
+├── tracers: (T, S)
+├── closure: Nothing
+├── buoyancy: SeawaterBuoyancy with g=9.80665 and BoussinesqEquationOfState{Float64} with ĝ = NegativeZDirection()
+└── coriolis: Nothing
+
+julia> GravitationalPotentialEnergy(model)
+KernelFunctionOperation at (Center, Center, Center)
+├── grid: 1×1×100 RectilinearGrid{Float64, Flat, Flat, Bounded} on CPU with 0×0×3 halo
+├── kernel_function: gravitational_potential_energy_ccc (generic function with 1 method)
+└── arguments: ("KernelFunctionOperation at (Center, Center, Center)", "KernelFunctionOperation at (Center, Center, Center)", "Tuple{Float64}")
+```
+
+To use a reference density pass the argument `reference_geopotential_height` to
+`GravitationalPotentialEnergy`:
+```jldoctest
+julia> using Oceananigans, SeawaterPolynomials.TEOS10;
+
+julia> using Oceanostics.DensityEquationTerms: GravitationalPotentialEnergy;
+
+julia> grid = RectilinearGrid(size=100, z=(-1000, 0), topology=(Flat, Flat, Bounded));
+
+julia> tracers = (:T, :S);
+
+julia> eos = TEOS10EquationOfState();
+
+julia> buoyancy = SeawaterBuoyancy(equation_of_state=eos);
+
+julia> model = NonhydrostaticModel(; grid, buoyancy, tracers);
+
+julia> reference_geopotential_height = 0; # density variable will be σ₀
+
+julia> GravitationalPotentialEnergy(model, reference_geopotential_height)
+KernelFunctionOperation at (Center, Center, Center)
+├── grid: 1×1×100 RectilinearGrid{Float64, Flat, Flat, Bounded} on CPU with 0×0×3 halo
+├── kernel_function: gravitational_potential_energy_ccc (generic function with 1 method)
+└── arguments: ("KernelFunctionOperation at (Center, Center, Center)", "KernelFunctionOperation at (Center, Center, Center)", "Tuple{Float64}")
+```
+"""
+@inline function GravitationalPotentialEnergy(model; location = (Center, Center, Center))
+
+    validate_location(location, "GravitationalPotentialEnergy")
+    validate_buoyancy(model.buoyancy)
+
+    grid = model.grid
+    ρ = seawater_density(model) # in-situ model density
+    Z = model_geopotential_height(model)
+    g = tuple(model.buoyancy.model.gravitational_acceleration)
+
+    return KernelFunctionOperation{Center, Center, Center}(gravitational_potential_energy_ccc, grid, ρ, Z, g)
+end
+@inline function GravitationalPotentialEnergy(model, reference_geopotential_height; location = (Center, Center, Center))
+
+    validate_location(location, "GravitationalPotentialEnergy")
+    validate_buoyancy(model.buoyancy)
+
+    grid = model.grid
+    σ = seawater_density(model, geopotential_height = reference_geopotential_height) # potential density reference to `reference_geopotential_height`
+    Z = model_geopotential_height(model)
+    g = tuple(model.buoyancy.model.gravitational_acceleration)
+
+    return KernelFunctionOperation{Center, Center, Center}(gravitational_potential_energy_ccc, grid, σ, Z, g)
+end
+
+@inline gravitational_potential_energy_ccc(i, j, k, grid, density, Z, g) = g[1] * density[i, j, k] * Z[i, j, k]
```

Comment:
Two comments here:

1. Should this be `g[3]` instead of `g[1]`?
2. In Oceanostics I'm trying to name small functions like these as the actual operation they perform. So I'd change the name from `gravitational_potential_energy_ccc()` to `gρz_ccc()` or something. Do you agree that it'd be more readable?

Reply:
Regarding 2. yes that is a great idea and I will update!

For 1.,

```julia
g = tuple(model.buoyancy.model.gravitational_acceleration)
```

is a `Tuple` with a single entry which is the value of `gravitational_acceleration`, so to access the scalar value I think it should be `g[1]`. Not very clean way but it was so I could get to work on `GPU`. Is there a more appropriate way to pass a scalar to a `KernelFunctionOperation`? Maybe use `const` or `local const`? Or perhaps I should rename to use a `NamedTuple` called `parameters`,

```julia
parameters = (g = model.buoyancy.model.gravitational_acceleration)
``` 

which more closely follows the current design in Oceanostics.jl?

Reply:
I think this is best solution

```julia
parameters = (g = model.buoyancy.model.gravitational_acceleration)
```

so have made the change.